lib/provisioning/ProvisionDispatcher.js

```diff
@@ -242,103 +281,111 @@ class ProvisionDispatcher {
+                                    { zkPath: `${this._zkEndpoint}${zkPath}`,
+                                      error: err });
+                    return;
+                }
+                this._owners = children;
+                this._log.debug('owners updated',
+                                { zkPath: `${this._zkEndpoint}${zkPath}`,
+                                  owners: this._owners });
+                if (this._isLeader) {
+                    this._redispatchProvisions();
+                }
+            })
+        );
     }
 
     _onMyselfChange(cb) {
         // monitor change in my content
         const myPath = `${OWNERS}/${this._myName}`;
-        this._client.getData(
-            myPath, event => {
-                this._log.debug('got owner self event',
-                    { zkPath: this._zkEndpoint + myPath,
-                        event });
-                this._onMyselfChange(cb);
-            },
-            (err, data) => {
-                if (err &&
-                    err.getCode() !== zookeeper.Exception.NO_NODE) {
-                    this._log.error('error in getData',
-                        { zkPath: this._zkEndpoint + myPath,
-                            error: err });
-                    return cb(err);
-                }
-                if (data !== undefined) {
-                    const provisionList = JSON.parse(data);
-                    this._log.info('provisioning update',
-                        { zkPath: this._zkEndpoint + myPath,
-                            provisionList });
-                    return cb(null, provisionList);
+        this._withRandomDelayIfTest(
+            () => this._client.getData(
+                myPath, event => {
+                    this._log.debug('got owner self event',
+                                    { zkPath: this._zkEndpoint + myPath,
+                                      event });
+                    this._onMyselfChange(cb);
+                },
+                (err, data) => {
+                    if (err &&
+                        err.getCode() !== zookeeper.Exception.NO_NODE) {
+                        this._log.error('error in getData',
+                                        { zkPath: this._zkEndpoint + myPath,
+                                          error: err });
+                        return cb(err);
+                    }
+                    if (data !== undefined) {
+                        const provisionList = JSON.parse(data);
+                        this._log.info('provisioning update',
+                                       { zkPath: this._zkEndpoint + myPath,
+                                         provisionList });
+                        return cb(null, provisionList);
+                    }
+                    return undefined;
                 }
-                return undefined;
-            }
+            )
         );
     }
 
     _onProvisionChange() {
```

Comment:
Why not add callbacks to all methods which are `async` instead of adding timers to manage race? I feel like all the race is happening because async methods are being treated as sync.

Reply:
Not clear for me what timer you're referring to (just in case the `_withRandomDelayIfTest` wrapper only introduces delays for testing purpose). In a nutshell, this is an event handler, not an async function, so there's no callback for this reason. This is called when zookeeper notifies us of a change in the nodes.

The race that I fixed was due to two events happening but the first event handler leading to call `_redispatchProvisions()` had to restart its processing due to a change coming from another event handler (`_onOwnersChange()`), but the first processing may be still writing stale state while the second one triggers, hence introducing a race condition. Making sure we retrigger `_redispatchProvisions` after the current one ends when new events occurred in the meantime solves this race.

Reply:
I meant methods like these `_onProvisionChange`, `_redispatchProvisions`, `_onLeaderChange`,`_onOwnerChange`, `_onProvisionChange`. All of these methods are asynchronous in nature and they are being called as if they are synchronous. This could lead to unmanageable race(?)

Reply:
All those functions are triggered from event handlers of zookeeper. (the `cb` argument of `_onMyselfChange()` is not a callback but an event handler function as well that can be called multiple times). So I don't see in which respect adding a callback argument to them would help managing race conditions, as any of those functions can be called at *any* time from zookeeper notifications. What we care about here should be taken care of, which is, making sure we redispatch all provisions *after* the latest notification that has an impact on the provisioning has been received, at which point what we redispatch is guaranteed to be the latest state.

Actually I started adding a callback to `_redispatchProvisions()`, so that I could embed it in to an `async.queue` to serialize all calls, but I realized this is extra work since we only want the leader to redispatch everything *once* after the latest event has been received, not once per every single event that has been received. So I removed the queue in favor of a set of flags to know if a redispatch is in progress or not.

Reply:
Approving this after our offline discussion.